DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/CourseSetup/ManageCourseControllerTests.cs

```diff
@@ -115,6 +116,193 @@ public void
             Assert.IsFalse(controller.ModelState.IsValid);
         }
 
+        [Test]
+        public void SaveCourseDetails_calls_correct_service_method_with_valid_inputs()
+        {
+            // Given
+            var model = new EditCourseDetailsViewModel(
+                101,
+                1,
+                "Name",
+                "v1",
+                false,
+                null,
+                false,
+                null,
+                true,
+                true,
+                true,
+                null,
+                null
+            );
+
+            A.CallTo(
+                () => courseService.UpdateCourseDetails(
+                    A<int>._,
+                    A<string>._,
+                    A<string>._,
+                    A<string>._,
+                    A<bool>._,
+                    A<int>._,
+                    A<int>._
+                )
+            ).DoesNothing();
+
+            // When
+            var result = controller.SaveCourseDetails(1, model);
+
+            // Then
+            A.CallTo(
+                () => courseService.UpdateCourseDetails(
+                    A<int>._,
+                    A<string>._,
+                    A<string>._,
+                    A<string>._,
+                    A<bool>._,
+                    A<int>._,
+                    A<int>._
+                )
+            ).MustHaveHappened();
+            result.Should().BeRedirectToActionResult().WithActionName("Index");
+        }
+
+        [Test]
+        public void
+            SaveCourseDetails_does_not_call_service_with_invalid_model()
+        {
+            // Given
+            var model = new EditCourseDetailsViewModel(
+                101,
+                1,
+                "Name",
+                "v1",
+                false,
+                "Password1234",
+                false,
+                "invalid email",
+                true,
+                true,
+                true,
+                "75",
+                "90"
+            );
+            controller.ModelState.AddModelError("Email", "Email address must not contain any whitespace characters");
+
+            // When
+            var result = controller.SaveCourseDetails(1, model);
+
+            // Then
+            A.CallTo(
+                () => courseService.UpdateCourseDetails(
+                    A<int>._,
+                    A<string>._,
+                    A<string>._,
+                    "invalid email",
+                    A<bool>._,
+                    A<int>._,
+                    A<int>._
+                )
+            ).MustNotHaveHappened();

```

Comment:
removed the explicit values, but the reason I put it in there was because the tests with MustNotHaveHappened for this call fail because the other two tests prior make this call. So it finds it among the calls twice. Do you know a way to handle that?

Reply:
Changed the set up so the tests don't interfere with each other now!